copycat-api/src/main/java/org/apache/kafka/copycat/connector/Connector.java

```diff
@@ -0,0 +1,121 @@
+ * tasks. Second, they are responsible for monitoring inputs for changes that require
+ * reconfiguration and notifying the Copycat runtime via the ConnectorContext. Continuing the
+ * previous example, the connector might periodically check for new tables and notify Copycat of
+ * additions and deletions. Copycat will then request new configurations and update the running
+ * Tasks.
+ * </p>
+ */
+public abstract class Connector {
+
+    protected ConnectorContext context;
+
+    /**
+     * Initialize this connector, using the provided ConnectorContext to notify the runtime of
+     * input configuration changes.
+     * @param ctx context object used to interact with the Copycat runtime
+     */
+    public void initialize(ConnectorContext ctx) {
+        context = ctx;
+    }
+
+    /**
+     * <p>
+     * Initialize this connector, using the provided ConnectorContext to notify the runtime of
+     * input configuration changes and using the provided set of Task configurations.
+     * This version is only used to recover from failures.
+     * </p>
+     * <p>
+     * The default implementation ignores the provided simply ignores the provided Task
+     * configurations. During recovery, Copycat will request an updated set of configurations and
+     * update the running Tasks appropriately. However, Connectors should implement special
+     * handling of this case if it will avoid unnecessary changes to running Tasks.
+     * </p>
+     *
+     * @param ctx context object used to interact with the Copycat runtime
+     * @param taskConfigs existing task configurations, which may be used when generating new task configs to avoid
+     *                    churn in partition to task assignments
+     */
+    public void initialize(ConnectorContext ctx, List<Properties> taskConfigs) {
+        context = ctx;
+        // Ignore taskConfigs. May result in more churn of tasks during recovery if updated configs
+        // are very different, but reduces the difficulty of implementing a Connector
+    }
+
+    /**
+     * Start this Connector. This method will only be called on a clean Connector, i.e. it has
+     * either just been instantiated and initialized or {@link #stop()} has been invoked.
+     *
+     * @param props configuration settings
+     * @throws CopycatException
+     */
+    public abstract void start(Properties props) throws CopycatException;
+
+    /**
+     * Reconfigure this Connector. Most implementations will not override this, using the default
+     * implementation that calls {@link #stop()} followed by {@link #start(Properties)}.
+     * Implementations only need to override this if they want to handle this process more
+     * efficiently, e.g. without shutting down network connections to the external system.
+     *
+     * @param props new configuration settings
+     * @throws CopycatException
+     */
+    public void reconfigure(Properties props) throws CopycatException {
```

Comment:
I like the idea of forcing connector developers to think about how upgrades will work when they need to change configurations. On the other hand, I'm not sure this will actually do anything -- people can easily just leave throw `UnsupportedOperationException` and ignore this whole process. I'm not convinced that adding these APIs actually helps since we aren't providing any patterns or code to help connector developers do this right, just adding API to tell them they should think about it.

That said, I don't feel strongly about this, so if people think it's worth it I can add those APIs.